src/space_usage/mod.rs

```diff
@@ -0,0 +1,283 @@
+    fn deserialize<D>(deserializer: D) -> Result<ByteCount, D::Error> where D: Deserializer<'de> {
+        Ok(ByteCount(usize::deserialize(deserializer)?))
+    }
+}
+
+impl Add for ByteCount {
+    type Output = ByteCount;
+    fn add(self, rhs: ByteCount) -> ByteCount {
+        ByteCount(self.0 + rhs.0)
+    }
+}
+
+impl AddAssign for ByteCount {
+    fn add_assign(&mut self, rhs: ByteCount) {
+        self.0 += rhs.0;
+    }
+}
+
+/// Represents combined space usage of an entire searcher and its component segments.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SearcherSpaceUsage {
+    segments: Vec<SegmentSpaceUsage>,
+    total: ByteCount,
+}
+
+impl SearcherSpaceUsage {
+    pub(crate) fn new() -> SearcherSpaceUsage {
+        SearcherSpaceUsage {
+            segments: Vec::new(),
+            total: ByteCount(0),
+        }
+    }
+
+    /// Add a segment, to `self`.
+    /// Performs no deduplication or other intelligence.
+    pub(crate) fn add_segment(&mut self, segment: SegmentSpaceUsage) {
+        self.total += segment.total();
+        self.segments.push(segment);
+    }
+
+    /// Per segment space usage
+    pub fn segments(&self) -> &[SegmentSpaceUsage] {
+        &self.segments[..]
+    }
+
+    /// Returns total byte usage of this searcher, including all large subcomponents.
+    /// Does not account for smaller things like `meta.json`.
+    pub fn total(&self) -> ByteCount {
+        self.total
+    }
+}
+
+/// Represents combined space usage for all of the large components comprising a segment.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct SegmentSpaceUsage {
+    num_docs: u32,
+
+    termdict: PerFieldSpaceUsage,
+    postings: PerFieldSpaceUsage,
+    positions: PerFieldSpaceUsage,
+    fast_fields: PerFieldSpaceUsage,
+    fieldnorms: PerFieldSpaceUsage,
+
+    store: StoreSpaceUsage,
+
+    deletes: ByteCount,
+
+    total: ByteCount,
+}
+
+impl SegmentSpaceUsage {
+    pub(crate) fn new(
+        num_docs: u32,
+        termdict: PerFieldSpaceUsage,
+        postings: PerFieldSpaceUsage,
+        positions: PerFieldSpaceUsage,
+        fast_fields: PerFieldSpaceUsage,
+        fieldnorms: PerFieldSpaceUsage,
+        store: StoreSpaceUsage,
+        deletes: ByteCount,
+    ) -> SegmentSpaceUsage {
+        let total = termdict.total()
+            + postings.total()
+            + positions.total()
+            + fast_fields.total()
+            + fieldnorms.total()
+            + store.total()
+            + deletes;
+        SegmentSpaceUsage {
+            num_docs,
+            termdict,
+            postings,
+            positions,
+            fast_fields,
+            fieldnorms,
+            store,
+            deletes,
+            total,
+        }
+    }
+
+    /// Num docs in segment
+    pub fn num_docs(&self) -> u32 {
+        self.num_docs
+    }
+
+    /// Space usage for term dictionary
+    pub fn termdict(&self) -> &PerFieldSpaceUsage {
```

Comment:
Could we replace all of the accessors by a `.get(&self, component: SegmentComponent) ` ?

Reply:
Not exactly, since `SegmentComponent::DELETE` and `SegmentComponent::STORE` are not per field things. So there would have to be an enum result type.

Reply:
Right.

Reply:
So I kind of went the opposite way. But it supports both use cases as well as makes updating almost certain to happen when a component is added.

Let me know if you feel strongly about deleting the other methods.

Reply:
ok...